Cohort Dropdown

dan-f opened the pull request:
[AN-6235](https://openedx.atlassian.net/secure/RapidBoard.jspa?rapidView=21&view=detail&selectedIssue=AN-6235)

@dsjen could you give this a review?

dsjen commented on .jshintrc:
```diff
@@ -142,6 +142,7 @@
         // Django translation
-        "gettext"
+        "gettext",
+        "ngettext"
```
I learned something new!

dsjen commented on analytics_dashboard/static/apps/learners/js/spec/roster-view-spec.js:
```diff
@@ -68,7 +71,10 @@ define([
         it('displays the last updated date', function () {
-            var roster = getRosterView({results: [{last_updated: new Date('1/1/2016')}]}, {parse: true});
+            var roster = getRosterView({
+                collectionResponse: {results: [{last_updated: new Date('1/1/2016')}]},
```
Nit: I tend to like to not use 1/1 as the date because it's the same result when the month and day are switched.

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/apps/learners/js/spec/roster-view-spec.js:
```diff
@@ -307,6 +319,60 @@ define([
+                            cohort: cohort
+                        }));
+                    } else {
+                        expect(getLastRequestParams().hasOwnProperty('cohort')).not.toBe(true);
```
Nit: `toBe(false)` would be more straightforward than `not.toBe(true)`

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/sass/_developer.scss:
```diff
@@ -776,4 +776,32 @@ table.dataTable thead th.sorting_desc:after {
+  }
+
+  .learners-cohort-filter {
+    padding-left: 20px;
```
It looks like the same `padding-left` value of `20px` is used here and in `learners-search`.  Moving this into a sass variable would allow us to adjust them a bit more easily.  Actually, it makes me think that another way to do this would be to nest the search and filter element in a div with the padding rather.  I don't have strong opinions on that front.

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/js/test/spec-runner.js:
```diff
@@ -34,6 +34,16 @@
+                if (count === 1) {
+                    return singularString;
+                } else {
+                    return pluralString
```
Missing a semicolon here.

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/apps/learners/templates/search.underscore:
```diff
@@ -1,7 +1,20 @@
+            type="search"
+            <% if (placeholder) { %> placeholder="<%- placeholder %>" <% } %> name="<%- name %>"
+        />
+        <a href="#" class="clear btn" data-backgrid-action="clear">
```
I believe this would benefit from `type=button`.

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/apps/learners/templates/search.underscore:
```diff
@@ -1,7 +1,20 @@
+        </a>
+    </div>
+    <span class="input-group-btn">
+        <a href="#" class="search btn btn-primary">
```
`type=button` here too.

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/apps/learners/templates/roster.underscore:
```diff
@@ -3,7 +3,9 @@
     <div class="col-md-4 col-md-push-8">
-        <div class="learners-search row"></div>
+        <div class="learners-search-container row"></div>
+        <hr>
```
I think the `<hr>` should be conditional if there are cohorts.  Otherwise, a divider appears and if we were to have more filters, there might be a traffic jam of dividers!

![screen shot 2016-03-04 at 3 33 51 pm](https://cloud.githubusercontent.com/assets/5265058/13539445/83653630-e21e-11e5-9bec-0e00987c9a18.png)

dan-f replied:
got this

dsjen commented on analytics_dashboard/static/apps/learners/templates/cohort-filter.underscore:
```diff
@@ -0,0 +1,11 @@
+    <%- selectDisplayName %>
+</label>
+<select id="cohort-filter" class="form-control">
+    <option selected value=""><%- noCohortSelectedMessage %></option>
```
Is it possible that there might be a cohort without a name (e.g. empty string)?  I seem to remember conversation about that, but didn't remember what the result was.  I think it was that wasn't valid.

dan-f replied:
Yeah the result of that investigation was that there are no cohorts with empty string names.

dsjen replied:
Excellent!

dsjen commented on analytics_dashboard/courses/templates/courses/learners.html:
```diff
@@ -21,7 +20,9 @@
              learnerListJson: {{ learner_list_json | escape_json }},
-             learnerListUrl: '{{ learner_list_url }}'
+             learnerListUrl: '{{ learner_list_url }}',
+             courseLearnerMetadataJson: {{ course_learner_metadata_json | escape_json }},
```
`escape_json` throws an error when `course_learner_metadata_json` is `None`:

```
AttributeError: 'NoneType' object has no attribute 'replace'
```

I check and test would be helpful!

dan-f replied:
Nice catch! I was surprised to see that we don't have view tests for this file. I'll add some tests.

dan-f replied:
Wrote the view tests, and will be pushing up soon.

dsjen commented on analytics_dashboard/static/apps/learners/js/app.js:
```diff
@@ -48,18 +58,26 @@ define([
-            var rootView = new LearnersRootView({el: $(this.options.containerSelector)}).render(),
-                router;
+            var router;
+            this.layout = new LearnersRootView({el: $(this.options.containerSelector)}).render();
```
Is there a need to save the layout?

dan-f replied:
I was trying to get the Chrome Marionette dev tools plugin to work, and it required this change. However that experiment didn't work, so I'll change it back.

dsjen commented on analytics_dashboard/static/apps/learners/js/models/course-metadata.js:
```diff
@@ -0,0 +1,28 @@
+define(['backbone'], function (Backbone) {
+    'use strict';
+
+    var CourseMetadataModel = Backbone.Model.extend({
```
Is there a spec/test for this model?  A simple one for the url would help to catch any inadvertent changes to the contract.

dan-f replied:
Thanks for the reminder. I was going to do that but... forgot :o

dan-f replied:
Got this. pushing up soon.

dsjen commented on analytics_dashboard/static/apps/learners/js/views/roster-view.js:
```diff
@@ -201,11 +210,72 @@ define([
+                    displayName: _.template(ngettext(
+                        // jshint ignore:start
+                        // Translators: 'cohortName' is the name of the cohort and 'numLearners' is the number of learners in that cohort.  The resulting phrase displays a cohort and the number of students belonging to it.
+                        '<%- cohortName %> (<%- numLearners %> learner)',
```
Numbers would be more readable if we localized the number.  https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/js/utils/utils.js#L58 can help.

dan-f replied:
Got this

dsjen commented on analytics_dashboard/static/apps/learners/js/views/roster-view.js:
```diff
@@ -201,11 +210,72 @@ define([
+                        // Translators: 'cohortName' is the name of the cohort and 'numLearners' is the number of learners in that cohort.  The resulting phrase displays a cohort and the number of students belonging to it.
+                        '<%- cohortName %> (<%- numLearners %> learner)',
+                        // Translators: 'cohortName' is the name of the cohort and 'numLearners' is the number of learners in that cohort.  The resulting phrase displays a cohort and the number of students belonging to it.
+                        '<%- cohortName %> (<%- numLearners %> learners)',
```
Same formatting for `numLearners` as above.

dan-f replied:
Got this

dsjen commented on analytics_dashboard/static/apps/learners/js/views/roster-view.js:
```diff
@@ -194,6 +197,12 @@ define([
      */
     LearnerSearch = Backgrid.Extension.ServerSideFilter.extend({
+        className: function () {
+            return Backgrid.Extension.ServerSideFilter.prototype.className + ' ' + 'learners-search';
```
Nit: Were you thinking to call out the extra space explicitly here?  It seems like this could be:
`Backgrid.Extension.ServerSideFilter.prototype.className + ' learners-search'` instead.

dan-f replied:
I did want to call out the extra space explicitly, since we're joining together class names. I'll rewrite it this way for clarity:

```javascript
return [Backgrid.Extension.ServerSideFilter.prototype.className, 'learners-search'].join(' ');
```

dsjen replied:
Beautiful! :+1:

dsjen commented on analytics_dashboard/static/apps/learners/js/views/roster-view.js:
```diff
@@ -201,11 +210,72 @@ define([
+                // Translators: "Cohort Groups" refers to groups of students within a course.
+                selectDisplayName: gettext('Cohort Groups'),
+                // Translators: "All" refers to viewing all the learners in a course.
+                noCohortSelectedMessage: gettext('All')
```
Nit: The name of this field is a bit confusing because it says no cohort, but then the message is "All."

dan-f replied:
Got this

dsjen commented on analytics_dashboard/static/apps/learners/js/spec/roster-view-spec.js:
```diff
@@ -307,6 +319,60 @@ define([
+
+                it('renders when the course contains cohorts', function () {
+                    var rosterView = getRosterView({courseMetadata: {cohorts: {
+                            'Cohort A': 1,
```
Is it possible that a cohort would have no learners?

dan-f replied:
Sure, but I don't see how that'd matter here. The API should return `{'Empty Cohort Name': 0}`, and we'll render "Empty Cohort Name (0 learners)".

dsjen commented:
Thanks for putting this together.  I've left a few comments; it's looking good!

dan-f commented:
@dsjen I addressed all your feedback. Please let me know if you're happy with my changes. I am in the process of adding a bit more HTTP error handling on the front end.

dan-f commented:
@dsjen I've pushed up a commit which adds more error handling to the front end. We should now be catching all kinds of errors (e.g. HTTP 500 as well as timeouts or network issues).

dsjen commented on analytics_dashboard/static/apps/learners/js/utils.js:
```diff
@@ -0,0 +1,28 @@
+         * @param jqXHR a jQuery XHR object.
+         * @param textStatus
+         */
+        handleAjaxFailure: function (jqXHR, textStatus) {
```
It might be nice for this to go in the general utils file, https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/js/utils.  It's up to you though as I guess the other pages won't use this in the foreseeable future.

dan-f replied:
Yeah I decided to put it here because I didn't want to prematurely pattern-ify this before it was tried and tested in the learners app. But I'd be happy to move it into the general utils file once we've vetted it a bit more.

dsjen commented:
Get job!  :+1:

dsjen commented:
Great job! :+1:
